sql: support ORDER BY / LIMIT clauses with DELETE and UPDATE statements

knz opened the pull request:
Suggested by @bdarnell in https://github.com/cockroachdb/cockroach/pull/18023#issuecomment-326049415:

> MySQL also extends DELETE with an ORDER BY clause to make the
> results of a LIMITed DELETE deterministic.

Actually this helps making the output of RETURNING deterministic in
general, whether or not LIMIT is present.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/18030)
<!-- Reviewable:end -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Since these are apparently so easy, I'll also mention that MySQL also adds `ORDER BY` and `LIMIT` to `UPDATE`. 

---

Review status: 0 of 7 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/sql/logictest/testdata/logic_test/delete, line 183 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18030#-Kso4dyXLR0pFsdSGjdl:-Kso4dyXLR0pFsdSGjdm:bz7wzeq) ([raw file](https://github.com/cockroachdb/cockroach/blob/896dd3f4a5db156e53c57f8d14d062962a9e7d52/pkg/sql/logictest/testdata/logic_test/delete#L183)):*
> ```
> 
> query ITTT
> EXPLAIN DELETE FROM unindexed WHERE v = 5 ORDER BY v LIMIT 10
> ```

Keep the old test to ensure that DELETE with LIMIT but without ORDER BY doesn't insert a `sort` step.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18030#-:-Kso50xnJqjd1BYT0qYX:bg9mkow)*
<!-- Sent from Reviewable.io -->

knz committed:
sql: support ORDER BY clauses with DELETE statements

Suggested by @bdarnell:

> MySQL also extends DELETE with an ORDER BY clause to make the
> results of a LIMITed DELETE deterministic.

Actually this helps making the output of RETURNING deterministic in
general, whether or not LIMIT is present.

knz commented:
Done, PTAL.

knz committed:
sql: support LIMIT and ORDER BY on UPDATE.

Suggested by @bdarnell for symmetry with DELETE.

The argument is similar: LIMIT can help control the size of a
transaction and ORDER BY enables deterministic results in RETURNING.
